### PR TITLE
Create synthesized record constructor

### PR DESCRIPTION
**Pull request by agocke**
This just adds a synthesized symbol (no body) for the constructor

### CODE REVIEW

**Comment by gafter** [comment]
https://github.com/dotnet/csharplang/blob/master/proposals/records-wip.md doesn't say anything about a constructor.

File: src/Compilers/CSharp/Test/Symbol/Symbols/Source/RecordTests.cs [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+            var comp = CreateCompilation(@"
+class C(int x, string y)
+{
+    public C(int a, int b) // overload
+    {
+    }
+}");
+            comp.VerifyDiagnostics();
+            var c = comp.GlobalNamespace.GetTypeMember("C");
+            var ctors = c.GetMembers(".ctor");
```
Comment (gafter) [inline src/Compilers/CSharp/Test/Symbol/Symbols/Source/RecordTests.cs]: Assert `ctors.Length`?

**Comment by agocke** [comment]
Spec work at https://github.com/dotnet/csharplang/pull/3076

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 2]
```diff
@@ -2841,8 +2846,73 @@ private void CheckForStructBadInitializers(MembersAndInitializersBuilder builder
...
+                {
+                    // PROTOTYPE: Add error for multiple parameter lists
+                }
+            }
+
+            if (paramList is null)
+            {
+                if (_declModifiers.HasFlag(DeclarationModifiers.Data))
+                {
+                    diagnostics.Add(ErrorCode.ERR_BadRecordDeclaration, declaration.NameLocations[0]);
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: >ERR_BadRecordDeclaration [](start = 46, length = 24)

This error is not supported by the draft specification.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 3]
```diff
@@ -2841,8 +2846,73 @@ private void CheckForStructBadInitializers(MembersAndInitializersBuilder builder
...
+                }
+                return;
+            }
+
+            if (!_declModifiers.HasFlag(DeclarationModifiers.Data))
+            {
+                diagnostics.Add(ErrorCode.ERR_BadRecordDeclaration, paramList.Location);
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: >ERR_BadRecordDeclaration [](start = 42, length = 24)

Same here.

**Comment by agocke** [comment]
@dotnet/roslyn-compiler  for one more review

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 4]
```diff
@@ -1,5 +1,7 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
+#nullable enable
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: In the future, please consider doing any nullable changes for a full file as a separate commit, so it's easy to see what's nullable annotation changes and the actual work being done in the PR.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 5]
```diff
@@ -199,7 +201,7 @@ public bool SetNullableContext(byte? value)
             _flags = new Flags(specialType, typeKind);
 
             var containingType = this.ContainingType;
-            if ((object)containingType != null && containingType.IsSealed && this.DeclaredAccessibility.HasProtected())
+            if (containingType?.IsSealed == true && this.DeclaredAccessibility.HasProtected())
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: Nit: Our general approach to nullable annotating has been to keep changes minimal. If you're going to go as far as to use `?.`, you may was well go all the way to pattern matching.

**Comment by 333fred** [comment]
>         internal override void ForceComplete(SourceLocation locationOpt, CancellationToken cancellationToken)

Should be annotated.

---
Refers to: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:429 in 2f48801. [](commit_id = 2f488019c7eab4ca705355d383e8c8bc2346240e, deletion_comment = False)

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs (hunk 5 above)
Comment (agocke) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: This is minimal only because the conversion was producing a nullable warning.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 6]
```diff
@@ -1528,7 +1531,7 @@ private void CheckMemberNameConflicts(DiagnosticBag diagnostics)
 
                     var conversion = symbol as SourceUserDefinedConversionSymbol;
                     var method = symbol as SourceMemberMethodSymbol;
-                    if ((object)conversion != null)
+                    if (!(conversion is null))
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: >!(conversion is null) [](start = 24, length = 21)

Consider `is object` instead.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 7]
```diff
@@ -1559,7 +1562,7 @@ private void CheckMemberNameConflicts(DiagnosticBag diagnostics)
                         // Do not add the conversion to the set of previously-seen methods; that set
                         // is only non-conversion methods.
                     }
-                    else if ((object)method != null)
+                    else if (!(method is null))
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: Same here.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 8]
```diff
@@ -2470,11 +2473,11 @@ internal Binder GetBinder(CSharpSyntaxNode syntaxNode)
                 foreach (SourceOrdinaryMethodSymbol method in methodsBySignature.Values)
                 {
                     // partial implementations not paired with a definition
-                    if (method.IsPartialImplementation && (object)method.OtherPartOfPartial == null)
+                    if (method.IsPartialImplementation && method.OtherPartOfPartial is null)
                     {
                         diagnostics.Add(ErrorCode.ERR_PartialMethodMustHaveLatent, method.Locations[0], method);
                     }
-                    else if ((object)method.OtherPartOfPartial != null && MemberSignatureComparer.ConsideringTupleNamesCreatesDifference(method, method.OtherPartOfPartial))
+                    else if (!(method.OtherPartOfPartial is null) && MemberSignatureComparer.ConsideringTupleNamesCreatesDifference(method, method.OtherPartOfPartial))
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: `is object`

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedRecordConstructor.cs [hunk 9]
```diff
@@ -0,0 +1,30 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedRecordConstructor.cs]: If you're going enable nullable elsewhere, please add it for new code.

**Comment by 333fred** [comment]
Done review pass (commit 4)

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs (hunk 6 above)
Comment (agocke) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: Never!

Comment (333fred) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: When you say never, you're being inconsistent with the rest of the compiler codebase.

Comment (gafter) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: I am not convinced by the assertion that `is object` is a clear and straightforward way to test for null.

Comment (gafter) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: I am not convinced by the assertion that `is object` is a clear and straightforward way to test for null.

Comment (agocke) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: There are plenty of places we don't use `is object` in the code base today. Until we settle on a single pattern, any of them are acceptable.

**Comment by 333fred** [comment]
@agocke still waiting for this to be addressed before approving: https://github.com/dotnet/roslyn/pull/40625#issuecomment-574888299

**Comment by agocke** [comment]
@333fred done

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 10]
```diff
@@ -1303,14 +1306,14 @@ private MembersAndInitializers GetMembersAndInitializers()
             AddDeclarationDiagnostics(diagnostics);
             diagnostics.Free();
 
-            return membersAndInitializers;
+            return _lazyMembersAndInitializers!;
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: >_lazyMembersAndInitializers [](start = 19, length = 27)

Was this change needed?

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 11]
```diff
@@ -2841,8 +2846,73 @@ private void CheckForStructBadInitializers(MembersAndInitializersBuilder builder
...
+                diagnostics.Add(ErrorCode.ERR_BadRecordDeclaration, paramList.Location);
+            }
+
+            BinderFactory binderFactory = this.DeclaringCompilation.GetBinderFactory(paramList.SyntaxTree);
+            var binder = binderFactory.GetBinder(paramList);
+
+            var memberSignatures = new HashSet<Symbol>(members, MemberSignatureComparer.DuplicateSourceComparer);
+
+            var ctor = new SynthesizedRecordConstructor(this, binder, paramList, diagnostics);
+            if (!memberSignatures.Contains(ctor))
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: Rather than constructing a `HashSet<>`, would it be more efficient to just call `members.Contains(ctor)`?

**Comment by cston** [comment]
>     [CompilerTrait(CompilerFeature.ReadOnlyReferences)]

Is this correct?

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:9 in 8d91657. [](commit_id = 8d916571a006ddd706957cb943160bac11691dd3, deletion_comment = False)

File: src/Compilers/CSharp/Test/Symbol/Symbols/Source/RecordTests.cs [hunk 12]
```diff
@@ -0,0 +1,114 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.CSharp.Test.Utilities;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.CSharp.UnitTests
+{
+    public class RecordTests : CompilingTestBase
```
Comment (cston) [inline src/Compilers/CSharp/Test/Symbol/Symbols/Source/RecordTests.cs]: >RecordTests [](start = 17, length = 11)

Why create a new class rather than adding to `Semantics.RecordTests`?

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs (hunk 11 above)
Comment (agocke) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: But I don't want to know if it has the exact symbol, I want to know if it has any symbol with a conflicting signature.

File: src/Compilers/CSharp/Test/Symbol/Symbols/Source/RecordTests.cs (hunk 12 above)
Comment (agocke) [reply src/Compilers/CSharp/Test/Symbol/Symbols/Source/RecordTests.cs]: Just splitting apart the checks of the symbol API from behavioral checks.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs (hunk 11 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: `members.Contains(ctor, MemberSignatureComparer.DuplicateSourceComparer)`

---
In reply to: [367670582](https://github.com/dotnet/roslyn/pull/40625#discussion_r367670582) [](ancestors = 367670582)

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs (hunk 10 above)
Comment (agocke) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: Probably not. I think I misread the early return there.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs (hunk 11 above)
Comment (agocke) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: Right, a linear walk. But this is just the ctor. I'm also gonna need to query for the other methods, like the generated properties, equals, hashcode, etc.
